File: validation/poc-00/java-candidate/src/main/java/dev/vra/poc00/domain/Money.java
package dev.vra.poc00.domain;

import java.math.BigDecimal;
import java.util.Currency;
import java.util.Objects;

/** Non-negative commerce value; not a signed finance/ledger model. */
public record Money(BigDecimal amount, Currency currency) {
    public Money {
        Objects.requireNonNull(amount, "amount");
        Objects.requireNonNull(currency, "currency");
        if (amount.signum() < 0) throw new IllegalArgumentException("Amount must be non-negative.");
        // Numeric equality must not depend on input scale (1.0 equals 1.00).
        amount = amount.stripTrailingZeros();
    }
    public Money add(Money other) {
        Objects.requireNonNull(other, "other");
        if (!currency.equals(other.currency)) throw new DomainFailure.CurrencyMismatch();
        return new Money(amount.add(other.amount), currency);
    }
}

File: validation/poc-00/kotlin-candidate/src/main/kotlin/dev/vra/poc00/kotlin/domain/Money.kt
package dev.vra.poc00.kotlin.domain

import java.math.BigDecimal
import java.util.Currency

/**
 * Non-negative exact commerce price/value, not a signed Finance/ledger amount.
 * Equality deliberately compares decimal value instead of BigDecimal scale.
 */
class Money(amount: BigDecimal, val currency: Currency) {
    val amount: BigDecimal = amount.stripTrailingZeros()

    init {
        require(this.amount.signum() >= 0) { "Amount must be non-negative." }
    }

    fun add(other: Money): Money {
        if (currency != other.currency) throw DomainFailure.CurrencyMismatch()
        return Money(amount.add(other.amount), currency)
    }

    override fun equals(other: Any?): Boolean =
        this === other || (other is Money && currency == other.currency && amount.compareTo(other.amount) == 0)

    override fun hashCode(): Int = 31 * currency.hashCode() + amount.stripTrailingZeros().hashCode()

    override fun toString(): String = "Money(amount=$amount, currency=$currency)"
}
